dynix.3.2.0/src/doc/man2/mmap.2
.\" $Copyright: $
.\" Copyright (c) 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991
.\" Sequent Computer Systems, Inc.   All rights reserved.
.\"  
.\" This software is furnished under a license and may be used
.\" only in accordance with the terms of that license and with the
.\" inclusion of the above copyright notice.   This software may not
.\" be provided or otherwise made available to, or used by, any
.\" other person.  No title to or ownership of the software is
.\" hereby transferred.
...
.V= $Header: mmap.2 1.30 1991/08/07 19:33:51 $
.TH MMAP 2 "\*(V)" "DYNIX" "\*(O)"
.SH NAME
mmap \- map an open file into the process's address space
.SH SYNOPSIS
.nf
.ft 3
#include <sys/types.h>
#include <sys/mman.h>

.ta 0.8i 1.6i
mmap(addr, len, prot, share, fd, pos)
caddr_t	addr;	/* starting virt-addr */
int	len;	/* length (bytes) to map */
int	prot;	/* RO, RW encoding */
int	share;	/* private/shared modifications */
int	fd;	/* open file to be mapped */
off_t	pos;	/* where in file to begin map */
.ft 1
.fi
.SH DESCRIPTION
.I Mmap
causes the file referenced by
.IR fd ,
starting at byte
.I pos
for
.I len
bytes in the file,
to be mapped into the calling
process's address space, starting at virtual address
.I addr
for
.I len
bytes, using protection specified by
.IR prot ;
modifications to the mapped memory are either
private to the process or
shared, as specified by
.IR share .
.I Mmap
can be used
to allocate regions of shared memory,
to map files into memory,
and to access special regions of the physical address space.
.PP
.I Fd
must reference an open regular (IFREG) or character special
(IFCHR) file.
The device driver that implements the IFCHR special file
must support mapping for this to succeed.
Typically, a regular file is used to map shared memory.
.PP
The
.I share
argument specifies whether modifications to a mapped file
are to be kept private to the calling process or shared with
other processes accessing the file.
If
.I share
is MAP_SHARED,
all modifications to the file are shared with others
who have it concurrently mapped.
If
.I share
is MAP_PRIVATE, all modifications are local to the calling process;
this doesn't restrict other processes from mapping the file.
MAP_SHARED and MAP_PRIVATE have no relation
to
.IR flock (2),
and do not restrict
.I read
and
.I write
system-calls.
If
.I share
is MAP_ZEROFILL, the space indicated from
.I addr
for
.I len
bytes is replaced by
private pages that are zero-filled when referenced.
MAP_ZEROFILL ignores the
.I fd
argument, and
.I pos
is ignored other than being checked for alignment (specifying fd and pos = 0
is recommended).
.PP
The
.I prot
argument should be PROT_WRITE for write access to the mapped region,
PROT_READ for read access,
PROT_EXEC for executable access.
These values can be ORed to obtain read-write access, etc.
For programming convenience, PROT_RDWR is defined as (PROT_READ|PROT_WRITE).
The file access permissions on
.I fd
must allow the requested access.
The
.I prot
argument affects only the calling process;
other processes mapping the same file may have different access.
.PP
The
.IR addr ,
.IR len ,
and
.I pos
arguments must be integral multiples of the system page size,
as defined by
.IR getpagesize (2).
It is possible to map over previously mapped pages.
If
.I addr
and
.I len
specify a nonexistent part of the process's address space,
the process's data segment is grown to accommodate the request,
and (for non parallel processes) the process ``break'' (see
.IR brk (2))
is set to the high end of the mapped region. When mmap is used with parallel
processes the value of ``break'' is unchanged and subtleties exists as 
explained in the "NOTES" section below.
Reference to any ``holes'' between the mapped region and
the rest of the data segment result in
a segmentation fault (SIGSEGV).
.I mmap
does allow mapping over text pages.  It does not
allow mapping over private stack pages.
.PP
When memory is mapped to a regular file,
the file acts like a paging area for the mapped memory region.
.I Read
and
.I write
operations to mapped regions of the file also affect the corresponding memory.
The memory contents are copied out to the file when the process is swapped
or when it exits, or when the region is otherwise unmapped by the last process
that has it mapped.
For programs that use shared memory but do not need a permanent disk image
of the memory,
the file associated with
.I fd
can be
.IR unlink ed
(see
.IR unlink (2))
even before the call to
.IR mmap :
if the file is
.IR unlink ed
when the region is unmapped,
the disk space will not be updated.
.PP
Regular files have their size rounded up to a file-system block boundary.
Any non-existent space in the file at the time of the
.I mmap
request (for example, in a sparse file) is allocated,
and filled with zeroes when referenced.
Both of these operations require write access to the file.
.PP
The type of file referenced may impose further restrictions on the
.IR pos ,
.IR offset ,
or other parameters.
Refer to the manual entry of the relevant device driver (for example,
.IR pmap (4))
for details.
.PP
.I Mmap
can be called multiple times with the same file descriptor,
resulting in several (possibly overlapping) mapped regions.
A process can have up to 8 regions mapped simultaneously;
mappings that are completely overlapped by subsequent mappings
are not counted in this total.
Mappings which use the same file descriptor, and
addr and pos arguments which align virtually with a previous mapping,
also don't count in this total; the simplest case is mapping more of a
file, starting from the end of a previous mapping.
.PP
All mapped files remain mapped in both the parent and child process after a
.IR fork .
All flavors of the
.I exec
and
.I exit
system calls, when successful,
remove all maps the calling process had established.
A child of a
.I vfork
that has not
yet
.IR exec ed
a new image cannot successfully execute
.IR mmap .
.PP
There are three types of mapping:
paged, physical, and non-paged memory.
The type of mapping is determined by the type of file being mapped.
Paged maps support shared memories and mapped regular files.
Physical maps deal with hardware that has restrictive access capability
(for example, the MULTIBUS address space, including Atomic Lock Memory).
Non-paged memory maps are typically used
for special reserved areas of system memory;
they are assumed to behave exactly like memory,
supporting accesses of arbitrary size and alignment, DMA, etc.
.PP
System services (raw IO, read/write, stat, etc.)
are supported in paged and non-paged memory maps;
attempts at such services in physically mapped address space
result in an error, typically EFAULT.
Core dumps include a copy of any mapped address space;
however, physically mapped addresses read as zero.
.PP
Regular files (IFREG) are always page-mapped.
Character special files (IFCHR) can support
paged, physical, or non-paged maps,
depending on the underlying hardware.
Physical and non-paged maps are
always valid in the process address space;
references won't cause a page fault.
.PP
When \f2mmap\fP increases a program's address
space, it also attempts to increase its allowable resident set size.
.SH "NOTES"
1) Interaction of mmap and libpps:
.PP
Parallel programs have a different virtual memory layout than regular
programs, and this effects the way mmap should be used. 
The virtual memory layout of the shared processes contains additional
regions of shared data, shared stack and shared heap. Any
mmapped region should be at an address well beyond the highest expected
value of the process's shared heap (see Guide to Parallel Programming).
The effective ``break'' value of a parallel
process may be obtained by calling the routine shsbrk(0),
instead of sbrk(0). The address to mmap should then be chosen well above
this value. Once mmap succeeds for  parallel program, the ``break'' value
for the system will remain unchanged, and it is up to the user's program
to keep track of the mmaped addresses.
.SH "RETURN VALUE"
.I Mmap
returns zero when successful.
Otherwise it returns -1 and places the error number in the global variable
.IR errno .
.SH EXAMPLES
The following code sets up a 1-Mbyte region of shared memory
at the first page boundary above the current program ``break.''
This region will be shared with the process's children
and with any other process that maps the file ``shmem''.
.ne 4
.Ps
pgsz = getpagesize();
shm_base = (char *) ( ((int)sbrk(0) + (pgsz-1)) & ~(pgsz-1) );
fd = open ("shmem", O_CREAT | O_RDWR, 0666);
mmap (shm_base, 0x100000, PROT_RDWR, MAP_SHARED, fd, 0);

Please note that with parallel programs the value of curshbrk 
is used instead of sbrk(0) as shown below:

pgsz = getpagesize();
shm_base = (char *) ( ((int)shsbrk(0) + (pgsz-1)) & ~(pgsz-1) );
fd = open ("shmem", O_CREAT | O_RDWR, 0666);
mmap (shm_base, 0x100000, PROT_RDWR, MAP_SHARED, fd, 0);
.Pe
.PP
The following code maps the first page of Atomic Lock Memory
into the process's virtual address space at address 0x200000.
This region will be shared with the process's children
and with any other process that maps the file ``/dev/alm/alm00''.
.Ps
pgsz = getpagesize();
fd = open ("/dev/alm/alm00", O_RDWR, 0);
mmap (0x200000, pgsz, PROT_RDWR, MAP_SHARED, fd, 0);
.Pe
.SH ERRORS
.TP 15
[EINVAL]
.IR Addr ,
.IR pos ,
or
.I len
is not a multiple of the system page size.
.TP 15
[EINVAL]
.I Prot
did not specify at least one of PROT_WRITE or PROT_READ;
.I share
did not specify MAP_SHARED, MAP_PRIVATE, or MAP_ZEROFILL; or
.I share
specified MAP_ZEROFILL but
.I prot
did not contain PROT_RDWR.
.TP 15
[EINVAL]
.I Fd
does not represent a regular or character special file.
.TP 15
[EINVAL]
The process is the child of a
.IR vfork .
.TP 15
[EINVAL]
The area defined by the
.I addr
and
.I len
arguments overlaps stack pages of the process.
.TP 15
[ENODEV]
The device driver indicated by
.I fd
does not support mapping.
.TP 15
[ENOMEM]
There is no swap space for the page table of a mapped regular file,
or you are trying to create too large a process.
.TP 15
[EMFILE]
The system-defined per-process limit on the number of
.IR mmap ed
files (currently 8)
was exceeded.
.TP 15
[ENFILE]
The system-wide limit on the number of mapped regular files was exceeded.
This limit is defined by the variable
.I nmfile
in
.IR /sys/conf/param.c .
.TP 15
[EACCES]
.I Fd
does not allow the desired access (read or write),
or a write-only file descriptor was used.
.TP 15
[EACCES]
A privately mapped file may not have the mapping specify
that the file should be extended; a private mapping is only
used to provide a fixed view of the file at the time the
mapping was established.
.TP 15
[EACCES]
A mapped regular file must be extended to a file-system block boundary,
or the file must have space allocated,
and the file descriptor is read-only.
.TP 15
[ENOSPC]
A mapped regular file was sparse
and there was insufficient space in the file-system to satisfy the request.
.TP 15
[EFBIG]
The
.I pos
and
.I len
arguments would create too large a file.
.TP 15
[others]
Other error values may be returned by some device drivers when requested to map.
See the relevant driver manual entry for details.
.SH "SEE ALSO"
munmap(2), pmap(4), vm_ctl(2), fork(2),
exec(2), getpagesize(2),
.I "Guide to Parallel Programming"
.SH "BUGS"
A mapped file may not be truncated.
.PP
If a file is extended to a file-system block boundary,
its original size is lost.
.PP
Current restrictions on what parts of the address space can be
re-mapped should be lifted.
.SH NOTES
Due to a hardware restriction, PROT_WRITE implies PROT_READ also.
.PP
To minimize overhead,
mapped regions should be kept as close as possible to the low end
of process memory.
.PP
Address space holes under the process ``break''
read as zeroes in core files.


dynix.3.2.0/src/usr.etc/snmpd/lib/mmap.c
/* $Copyright: $
 * Copyright (c) 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991
 * Sequent Computer Systems, Inc.   All rights reserved.
 *  
 * This software is furnished under a license and may be used
 * only in accordance with the terms of that license and with the
 * inclusion of the above copyright notice.   This software may not
 * be provided or otherwise made available to, or used by, any
 * other person.  No title to or ownership of the software is
 * hereby transferred.

/* $Header: mmap.c 1.2 1991/08/02 16:12:09 $ */

/* $Log: mmap.c,v $
 *
 *
 */

/* mmap.c:
 *
 * As a quick and easy way to get at things inside of the D3 kernel, we
 * mmap a lot of stuff.  For example, the process of retrieving all the
 * connections implies walking arbitrary mbuf memory.  So we mmap the
 * whole address space and walk it from user space (as opposed to implementing
 * such a feature in the kernel).  Although less reliable than a
 * kernel implementation, this is considerably lower drag, quicker
 * to implement, and keeps the kernel size small.  Besides, snmp is
 * a best effort protocol anyway.
 *
 * The mmap'd regions are managed kind of wierdly due to the system limitation
 * that we have at most X mmap'd regions (3.2: 16).  Since the daemon
 * exceeds that number, we take all the variable from the data and bss
 * areas, and map the area which is collectively that from the lowest
 * addressed variable to the highest addressed variable plus its length.
 * In other words, we mmap one huge region for all the data/bss symbols.
 * Mbufs, on the other hand, are calloc'd during system initialization (from
 * high memory, no less), so we mmap that as a separate region to prevent
 * this process form hogging up to many pte's.
 *
 * Other variable are read from /dev/kmem once at startup and are simply 
 * cached in this process for its duration; these presumably don't 
 * change anyway.
 */

#include "defs.h"

#include <nlist.h>
#include <sys/mman.h>
#include <sys/file.h>

extern	char *sbrk();
extern	char *valloc();
int	pagesize;
char	*unix_file = "/dynix";
char	*kmem_file = "/dev/kmem";
int	km;

/*
 * Globals which are initialized once.
 */
int	arptab_size;
int	tcp_backoff[2];
off_t	mdiff;		/* difference between our VM addresses and kernels */
int	ctl_sock;
int	rthashsize;
int	*ip_ttl;
struct	inpcb *tcb_addrs[INP_HASHSZ];
struct	inpcb *udb_addrs[INP_HASHSZ];
off_t	ifnet;

/*
 * Mbuf memory is not in bss, but in its own address space somewhere in
 * high memory.  So we manage it a separate region
 */
off_t	mblop, mbhip;
struct	mbuf *mlo, *mhi;

/*
 * Globals in data/bss which are memory mapped.  We map one large region,
 * then set these pointers up as offsets into that region
 */
struct	arptab *arptab;
struct  icmpstat *icmpstat;
struct	tcpstat *tcpstat;
struct	ipstat *ipstat;
struct	udpstat *udpstat;
struct	inpcb *tcb;
struct	inpcb *udb;
int	*ipforwarding;
struct	ifqueue	*ipintrq;;
struct	mbuf **rthost;
struct	mbuf **rtnet;


struct	nlist nl[] = {
	{ "_arptab_size" },
#define	ARPTAB_SIZE	0
	{ "_arptab" },
#define	ARPTAB		1
	{ "_icmpstat" },
#define	ICMPSTAT	2
	{ "_tcp_backoff" },
#define	TCP_BACKOFF	3
	{ "_tcpstat" },
#define	TCPSTAT		4
	{ "_tcb" },
#define	TCB		5
	{ "_mblorange" },
#define	MBLORANGE	6
	{ "_mbhirange" },
#define	MBHIRANGE	7
	{ "_udb" },
#define	UDB		8
	{ "_ifnet" },
#define	IFNET		9
	{ "_ipstat" },
#define	IPSTAT		10
	{ "_ipforwarding" },
#define	IPFORWARDING	11
	{ "_ipintrq" },
#define	IPINTRQ		12
	{ "_rthost" },
#define	RTHOST		13
	{ "_rtnet" },
#define	RTNET		14
	{ "_rthashsize" },
#define	RTHASHSIZE	15
	{ "_udpstat" },
#define	UDPSTAT		16
#ifdef	KERN3_2
	{ "_ip_ttl" },
#define	IP_TTL		17
#define	NLIST_MAX	18
#else
#define	NLIST_MAX	17
#endif
	0
};

/* 
 * init_mmanage:
 *
 * Manage the mmap'd address space which covers all of data/bss.  The
 * mm_dontcare field is used to indicate those variables which aren't
 * part of the mmap.  The mm_offset field is used to indicate the
 * symbols's offset in kernel virtual memory, and mm_len is the
 * variable length.  Later this data structure is used to calculate
 * the address range needed to cover all the data/bss symbols.
 */

struct mmanage {
	off_t	mm_offset;
	int	mm_len;
	int	mm_dontcare;
}mm[NLIST_MAX];

init_mmanage()
{
	int i;

	for (i = 0; i < NLIST_MAX; i++)
		mm[i].mm_offset = nl[i].n_value;
	mm[ARPTAB_SIZE].mm_dontcare = 1;
	mm[TCP_BACKOFF].mm_dontcare = 1;
	mm[MBLORANGE].mm_dontcare = 1;
	mm[MBHIRANGE].mm_dontcare = 1;
	mm[RTHASHSIZE].mm_dontcare = 1;
	mm[IFNET].mm_dontcare = 1;

	mm[ARPTAB].mm_len = sizeof(struct arptab) * arptab_size;
	mm[ICMPSTAT].mm_len = sizeof(struct icmpstat);
	mm[TCPSTAT].mm_len = sizeof(tcpstat);
	mm[TCB].mm_len = sizeof(struct inpcb) * INP_HASHSZ;
	mm[UDB].mm_len = sizeof(struct inpcb) * INP_HASHSZ;
	mm[IPSTAT].mm_len = sizeof(struct ipstat);
	mm[IPFORWARDING].mm_len = sizeof(int);
	mm[IPINTRQ].mm_len = sizeof(struct ifqueue);
	mm[RTHOST].mm_len = sizeof(struct mbuf *) * rthashsize;
	mm[RTNET].mm_len = sizeof(struct mbuf *) * rthashsize;
	mm[UDPSTAT].mm_len = sizeof(struct udpstat);

#ifdef	KERN3_2
	mm[IP_TTL].mm_len = sizeof(int);
#endif
}

/*
 * do_mmap:
 *
 * Actually mmap a kernel virtual address into our own process address
 * space.
 */

char*
do_mmap(size, kernaddr)
	int size;
	u_long kernaddr;
{
        off_t	pos;
        int	sz, off;
        char	*va;
	static	times = 0;
	int 	i;

	va = (caddr_t) (((int)sbrk(0) + (pagesize-1)) & ~(pagesize-1));
        pos = (unsigned)kernaddr & ~(pagesize-1);
        off = (unsigned)kernaddr - pos;
        sz = size + off;
        sz = (sz+pagesize-1) & ~(pagesize-1);
	if (mmap(va, sz, PROT_RDWR, MAP_SHARED, km, pos) != 0) {
		syslog(LOG_ERR, "mmap: %m");
		exit(1);
	}
        return((va + off));
}
/*
 * find_region:
 * 
 * find the lowest virtual address out of all symbols we wish to map
 * At the same time, find the highest address we need to mmap
 */
find_region(lp, hp)
	u_long *lp;
	u_long *hp;
{
	int i;
	u_long min = 0xffffffff;
	u_long max = 0;

	for (i = 0; i < NLIST_MAX; i++) {
		if (mm[i].mm_dontcare == 1)
			continue;
		min = MIN(min, mm[i].mm_offset);
		max = MAX(max, mm[i].mm_offset + mm[i].mm_len);
	}

	*lp = min;
	*hp = max;
}

/* 
 * init_mmap:
 *
 * Initialize all mmap's.  This is called from all the get routines, so
 * the daemon startup time is negligible -- the first query takes the
 * hit for setting everything up.  At the same time the mmap's are
 * initialized, those variable which only need be read and not mmap'd
 * are also set up.
 */
init_mmap()
{
	static int initted = 0;
	int error = 0;
	int i, size;
	char *start;
	u_long max, base;

	if (initted)
		return;
	initted = 1;

	if ((km = open(kmem_file, O_RDWR)) < 0) {
		fprintf(stderr, "Can't open kmem file %s\n", kmem_file);
		syslog(LOG_ERR, "open: %m");
		exit(1);
	}

	nlist(unix_file, nl);

	for (i = 0; i < sizeof(nl)/sizeof(struct nlist) - 1; i++) {
		if (nl[i].n_value == 0) {
			syslog(LOG_ERR, "Symbol '%s' not found in %s\n",
				nl[i].n_name, unix_file);
			error = 1;
		}
	}
	if (error)
		exit(1);

	pagesize = getpagesize();
	lseek(km, nl[ARPTAB_SIZE].n_value, 0);
	if (read(km, &arptab_size, sizeof(arptab_size)) 
	    != sizeof(arptab_size)) {
		syslog(LOG_ERR, "read arptab_size: %m");
		exit(1);
	}
	lseek(km, nl[MBLORANGE].n_value, 0);
	if (read(km, &mblop, sizeof(mblop)) != sizeof(mblop)) {
		syslog(LOG_ERR, "read mblop: %m");
		exit(1);
	}
	lseek(km, nl[MBHIRANGE].n_value, 0);
	if (read(km, &mbhip, sizeof(mbhip)) != sizeof(mbhip)) {
		syslog(LOG_ERR, "read mbhip: %m");
		exit(1);
	}
	lseek(km, nl[TCP_BACKOFF].n_value, 0);
	if (read(km, &tcp_backoff[0], sizeof(int)) != sizeof(int)) {
		syslog(LOG_ERR, "read tcp_backoff: %m");
		exit(1);
	}
	lseek(km, nl[TCP_BACKOFF].n_value + sizeof(int) * (TCP_MAXRXTSHIFT - 1),
		0);
	if (read(km, &tcp_backoff[1], sizeof(int)) != sizeof(int)) {
		syslog(LOG_ERR, "read tcp_backoff: %m");
		exit(1);
	}

	lseek(km, nl[RTHASHSIZE].n_value, 0);
	if (read(km, &rthashsize, sizeof(rthashsize)) != sizeof(rthashsize)) {
		syslog(LOG_ERR, "read rthashsize: %m");
		exit(1);
	}
	ifnet = nl[IFNET].n_value;

	size = (int) mbhip - (int) mblop;
	size = (size + (CLBYTES-1)) &~ (CLBYTES-1);
	mlo = (struct mbuf *) do_mmap(size, mblop);
	mhi = (struct mbuf *) ((int)mlo + size);
	mdiff = (off_t) ((caddr_t) mblop - (caddr_t) mlo);

	if ((ctl_sock = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
		syslog(LOG_ERR, "get interface configuration socket: %m");
		exit(1);
	}

	init_mmanage();
	find_region(&base, &max);

	start = do_mmap(max - base, base);
	init_pointers(base, start);

}

/*
 * init_pcbs
 *
 * Initialize the hash head pointers for the inpcb address lists.  The
 * The inpcbs are doubly threaded, and point back to the head of the
 * list.  So list searches terminate when inp_next == head address.
 * The idea here is that we have a "phony" hash list which only contains
 * the kernel address of the head record, so we can terminate list
 * searches.
 */
init_pcbs(head, off)
	struct inpcb *head[];
	off_t off;
{
	int i;

	for (i = 0; i < INP_HASHSZ; i++) {
		head[i] = (struct inpcb *)off;
		off += sizeof(struct inpcb);
	}
}

/*
 * init_pointers
 *
 * Initialize pointers to mmap'd regions.  Once done, when dereferenced,
 * we're talking directly to kernel virtual memory.
 */
init_pointers(base, start)
	u_long base;	/* lowest virtual address */
	char *start;	/* starting virtual address */

{
	arptab = (struct arptab *)(start + (nl[ARPTAB].n_value - base));
	icmpstat = (struct icmpstat *)(start + (nl[ICMPSTAT].n_value - base));
	tcpstat = (struct tcpstat *)(start + (nl[TCPSTAT].n_value - base));
	ipstat = (struct ipstat *)(start + (nl[IPSTAT].n_value - base));
	udpstat = (struct udpstat *)(start + (nl[UDPSTAT].n_value - base));
	tcb = (struct inpcb *)(start + (nl[TCB].n_value - base));
	udb = (struct inpcb *)(start + (nl[UDB].n_value - base));
	ipforwarding = (int *)(start + (nl[IPFORWARDING].n_value - base));
	ipintrq = (struct ifqueue *)(start + (nl[IPINTRQ].n_value - base));
	rthost = (struct mbuf **)(start + (nl[RTHOST].n_value - base));
	rtnet = (struct mbuf **)(start + (nl[RTNET].n_value - base));
#ifdef	KERN3_2
	ip_ttl = (int *)(start + (nl[IP_TTL].n_value - base));
#endif
}
